processor/src/main/java/org/mapstruct/ap/internal/model/ForgedMethod.java

```diff
@@ -347,7 +348,7 @@ public boolean isObjectFactory() {
 
     @Override
     public MappingMethodOptions getOptions() {
-        return basedOn.getOptions();
+        return MappingMethodOptions.getForgedMethodInheritedOptions( basedOn.getOptions() );
```

Comment:
Why is this needed? Is this actually the fix for #2438? I would prefer if we can have that separate to reduce the surface area of a single PR.

Reply:
I think this change was to prevent infinite loops, so that the `@SubclassMapping` is not inherited. Will double-check this to be certain.

Reply:
Makes sense, you need to remove the inherited `SubClassMapping`(s).  Let's create a field for this instead of creating it on every `getOptions` call. Those getOptions are called in a lot of places, so it might cause some performance problem.